quickwit-storage/src/storage_with_upload_cache.rs

```diff
@@ -0,0 +1,630 @@
+// Copyright (C) 2021 Quickwit, Inc.
+//
+// Quickwit is offered under the AGPL v3.0 and as commercial software.
+// For commercial licensing, contact us at hello@quickwit.io.
+//
+// AGPL:
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program. If not, see <http://www.gnu.org/licenses/>.
+
+use std::collections::HashMap;
+use std::fs;
+use std::ops::Range;
+use std::path::{Path, PathBuf};
+use std::sync::Arc;
+
+use async_trait::async_trait;
+use bytes::Bytes;
+use serde::{Deserialize, Serialize};
+use tokio::sync::Mutex;
+use tracing::warn;
+
+use crate::{LocalFileStorage, PutPayload, Storage, StorageErrorKind, StorageResult};
+
+// TODO(asking): could we make it hidden with `.quickwit-cache`
+const INTERNAL_CACHE_DIR_NAME: &str = "quickwit-cache";
+const CACHE_TEMP_FILE_EXTENSION: &str = "_quickwit_cache_temp_file";
```

Comment:
I don't know which files are considered as `quickwit cache temp` files, can you give an example to better understand this case?

Reply:
this is a work in progress, I still need to est and add comment